### PR TITLE
Grpc reconnect on disconnect

### PR DESCRIPTION
**Pull request by WilfredAlmeida**

### CODE REVIEW

File: client/main.rs [hunk 1]
```diff
@@ -96,3 +105,33 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
     info!("Service stopped.");
     Ok(())
 }
+
+async fn with_retry(
```
Comment (linuskendall) [inline client/main.rs]: rename to spawn_grpc_connection_with_retry

with_retry is a bit too generic

File: client/grpc_client.rs [hunk 2]
```diff
@@ -97,11 +97,13 @@ async fn mtx_stream(
...
                     error!("Upstream closed!");
-                    break
+                    // break
+                    return Err("Upstream closed!".into());
                 }
             }
         }
-    }
+    };
+    return Ok(())
```
Comment (fanatid) [inline client/grpc_client.rs]:
```suggestion
    Ok(())
```

File: client/grpc_client.rs [hunk 3]
```diff
@@ -158,13 +160,22 @@ pub async fn spawn_grpc_client(
...
+    match mx {
+        Ok(_) => {
+            info!("Stream ended successfully");
+        }
+        Err(e) => {
+            return Err(e);
+        }
+    }
+
     Ok(())
```
Comment (fanatid) [inline client/grpc_client.rs]:
```rust
    mtx_stream(
        grpc_host.unwrap_or("unknown").to_string(),
        &mut client,
        tx_transactions,
        metrics,
    )
    .await
    .inspect(|_| info!("Stream ended successfully"))
```

**Review by fanatid — COMMENTED** [review]
lgtm

File: client/grpc_client.rs (hunk 3 above)
Comment (WilfredAlmeida) [reply client/grpc_client.rs]: This throws `use of unstable library feature 'result_option_inspect'` for the `.inspect()` call

I'd say let's not use anything unstable. The whole thing is unstable enough already

Comment (fanatid) [reply client/grpc_client.rs]: ah, this is in the stable from 1.76.0 https://doc.rust-lang.org/std/result/enum.Result.html#method.inspect
I think we are okay to bump rust to 1.76.0 (in separate PR)

Comment (fanatid) [reply client/grpc_client.rs]: if we want to stay with 1.70.0 we can use map
```rust
.map(|v| {
  info!("Stream ended successfully");
  v
})
```

**Review by linuskendall — APPROVED** [review]

File: client/grpc_client.rs [hunk 4]
```diff
@@ -83,10 +83,11 @@ async fn mtx_stream(
                 if let Some(metrics) = metrics {
                     info!("Sending metrics: {:?}", metrics);
                     if let Err(err) = tx_upstream_transactions.send(metrics) {
-                        error!("Failed to enqueue metrics: {}", err);
+                        error!("Failed to enqueue metrics, source: {:?} err {}", source, err);
                     }
                 } else {
-                    error!("Stream of metrics dropped!");
+                    error!("Stream of metrics dropped: {:?}", source);
+                    break
```
Comment (fanatid) [inline client/grpc_client.rs:90]: double break :)
I think error would happen here because first one requires `;`

**Pull request merged**